File: app/evidence/engine.js
/**
 * Evidence Engine
 * Module 5: Generates and manages evidence for compliance results
 * Every compliance result must have evidence - no bare PASS/FAIL
 */

class EvidenceEngine {
    constructor(config = {}) {
        this.config = config;
        this.evidenceStore = new Map(); // In production, use database
    }

    /**
     * Create evidence record from compliance evaluation
     * @param {object} params
     * @returns {object} Evidence record
     */
    createEvidence(params) {
        const {
            requirement_id,
            document_id,
            page_number,
            extracted_value,
            expected_value,
            actual_value,
            verification_source,
            rule_id,
            result,
            confidence,
            details = {}
        } = params;

        const evidence = {
            id: this.generateId(),
            requirement_id,
            document_id,
            page_number: page_number || 1,
            extracted_value: String(extracted_value),
            expected_value: String(expected_value),
            actual_value: String(actual_value),
            verification_source,
            rule_id,
            result,
            confidence: Number(confidence) || 0,
            details,
            created_at: new Date().toISOString(),
            hash: this.generateHash(`${requirement_id}-${document_id}-${extracted_value}`)
        };

        this.evidenceStore.set(evidence.id, evidence);
        return evidence;
    }

    /**
     * Create multiple evidence records from rule results
     * @param {Array} ruleResults
     * @param {object} verifications
     * @param {object} fields
     * @param {Array} documents
     * @param {string} bidId
     * @returns {Array}
     */
    createEvidenceFromRules(ruleResults, verifications, fields, documents, bidId = null) {
        return ruleResults
            .filter(r => r.evidence_ref)
            .map(rule => this.createEvidence({
                requirement_id: rule.rule_id,
                document_id: rule.evidence_ref,
                page_number: this.getPageNumber(rule, fields),
                extracted_value: rule.actual,
                expected_value: rule.expected,
                actual_value: rule.actual,
                verification_source: rule.source || rule.rule_type,
                rule_id: rule.rule_id,
                result: rule.result,
                confidence: rule.confidence,
                details: { ...rule.details, bid_id: bidId }
            }));
    }

    /**
     * Get evidence by ID
     * @param {string} evidenceId
     * @returns {object|null}
     */
    getEvidence(evidenceId) {
        return this.evidenceStore.get(evidenceId) || null;
    }

    /**
     * Get all evidence for a bid
     * @param {string} bidId
     * @returns {Array}
     */
    getEvidenceForBid(bidId) {
        return Array.from(this.evidenceStore.values())
            .filter(e => e.details?.bid_id === bidId || e.requirement_id?.includes(bidId));
    }

    /**
     * Get evidence by requirement ID
     * @param {string} requirementId
     * @returns {Array}
     */
    getEvidenceByRequirement(requirementId) {
        return Array.from(this.evidenceStore.values())
            .filter(e => e.requirement_id === requirementId);
    }

    /**
     * Get evidence by document ID
     * @param {string} documentId
     * @returns {Array}
     */
    getEvidenceByDocument(documentId) {
        return Array.from(this.evidenceStore.values())
            .filter(e => e.document_id === documentId);
    }

    /**
     * Search evidence
     * @param {object} query
     * @returns {Array}
     */
    searchEvidence(query = {}) {
        let results = Array.from(this.evidenceStore.values());

        if (query.result) {
            results = results.filter(e => e.result === query.result);
        }
        if (query.verification_source) {
            results = results.filter(e => e.verification_source === query.verification_source);
        }
        if (query.rule_id) {
            results = results.filter(e => e.rule_id === query.rule_id);
        }
        if (query.min_confidence) {
            results = results.filter(e => e.confidence >= query.min_confidence);
        }
        if (query.bid_id) {
            results = results.filter(e => e.details?.bid_id === query.bid_id);
        }

        return results.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
    }

    /**
     * Add manual override evidence
     * @param {object} params
     * @returns {object}
     */
    addOverrideEvidence(params) {
        const {
            officer_id,
            original_result,
            new_result,
            reason,
            requirement_id,
            rule_id
        } = params;

        const evidence = this.createEvidence({
            requirement_id,
            document_id: `OVERRIDE-${Date.now()}`,
            page_number: 0,
            extracted_value: original_result,
            expected_value: original_result,
            actual_value: new_result,
            verification_source: 'MANUAL_OVERRIDE',
            rule_id,
            result: new_result,
            confidence: 1.0,
            details: {
                officer_id,
                original_result,
                new_result,
                reason,
                override: true,
                timestamp: new Date().toISOString()
            }
        });

        return evidence;
    }

    /**
     * Get evidence summary for a bid
     * @param {string} bidId
     * @returns {object}
     */
    getEvidenceSummary(bidId) {
        const evidence = this.getEvidenceForBid(bidId);

        const byResult = {};
        const bySource = {};

        for (const e of evidence) {
            byResult[e.result] = (byResult[e.result] || 0) + 1;
            bySource[e.verification_source] = (bySource[e.verification_source] || 0) + 1;
        }

        return {
            bid_id: bidId,
            total_evidence: evidence.length,
            by_result: byResult,
            by_source: bySource,
            evidence: evidence
        };
    }

    /**
     * Export evidence for audit trail
     * @param {string} bidId
     * @returns {object}
     */
    exportForAudit(bidId) {
        const evidence = this.getEvidenceForBid(bidId);

        return {
            bid_id: bidId,
            exported_at: new Date().toISOString(),
            total_records: evidence.length,
            records: evidence.map(e => ({
                requirement_id: e.requirement_id,
                rule_id: e.rule_id,
                result: e.result,
                expected: e.expected_value,
                actual: e.actual_value,
                source: e.verification_source,
                confidence: e.confidence,
                document: e.document_id,
                page: e.page_number,
                details: e.details,
                hash: e.hash,
                created_at: e.created_at
            }))
        };
    }

    // Helper methods
    generateId() {
        return 'EVD-' + Date.now().toString(36) + '-' + Math.random().toString(36).substr(2, 9);
    }

    generateHash(data) {
        let hash = 0;
        for (let i = 0; i < data.length; i++) {
            hash = ((hash << 5) - hash) + data.charCodeAt(i);
            hash |= 0;
        }
        return 'sha256:' + Math.abs(hash).toString(16).padStart(64, '0');
    }

    getPageNumber(rule, fields) {
        if (rule.rule_id === 'REQ-LC-01') return fields.local_content_source_page || 1;
        if (rule.rule_id === 'REQ-OEM-01') return fields.oem_source_page || 1;
        if (rule.details?.source_page) return rule.details.source_page;
        return 1;
    }

    /**
     * Clear evidence store (for testing)
     */
    clear() {
        this.evidenceStore.clear();
    }
}

module.exports = EvidenceEngine;

File: backend/app/evidence/engine.py
from typing import Dict, Any, List, Optional
from datetime import datetime, timezone
import hashlib


class EvidenceRecord:
    def __init__(self, **kwargs):
        self.id = kwargs.get("id", self._generate_id())
        self.requirement_id = kwargs.get("requirement_id", "")
        self.document_id = kwargs.get("document_id", "")
        self.page_number = kwargs.get("page_number", 1)
        self.extracted_value = str(kwargs.get("extracted_value", ""))
        self.expected_value = str(kwargs.get("expected_value", ""))
        self.actual_value = str(kwargs.get("actual_value", ""))
        self.verification_source = kwargs.get("verification_source", "")
        self.rule_id = kwargs.get("rule_id", "")
        self.result = kwargs.get("result", "")
        self.confidence = float(kwargs.get("confidence", 0))
        self.details = kwargs.get("details", {})
        self.created_at = kwargs.get("created_at", datetime.now(timezone.utc).isoformat())
        self.hash = kwargs.get("hash", self._generate_hash())

    def _generate_id(self) -> str:
        return f"EVD-{int(datetime.now().timestamp() * 1000)}-{hashlib.md5(str(datetime.now()).encode()).hexdigest()[:9]}"

    def _generate_hash(self) -> str:
        data = f"{self.requirement_id}-{self.document_id}-{self.extracted_value}"
        return "sha256:" + hashlib.sha256(data.encode()).hexdigest()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "requirement_id": self.requirement_id,
            "document_id": self.document_id,
            "page_number": self.page_number,
            "extracted_value": self.extracted_value,
            "expected_value": self.expected_value,
            "actual_value": self.actual_value,
            "verification_source": self.verification_source,
            "rule_id": self.rule_id,
            "result": self.result,
            "confidence": self.confidence,
            "details": self.details,
            "created_at": self.created_at,
            "hash": self.hash
        }


class EvidenceEngine:
    """In-memory evidence store.

    The canonical evidence record for a tender/bid lives in the ``documents``
    table as JSON (``Document.ai_extracted``).  This in-memory store is used
    only for the compliance and evidence API routes that return structured
    evidence objects.  When the server restarts the in-memory store is cleared.

    TODO: replace with a dedicated ``evidence_records`` DB table so that
    evidence survives restarts and can be queried with SQL joins.
    """

    def __init__(self):
        self.evidence_store: Dict[str, EvidenceRecord] = {}

    def create_evidence(self, params: Dict[str, Any]) -> EvidenceRecord:
        evidence = EvidenceRecord(**params)
        self.evidence_store[evidence.id] = evidence
        return evidence

    def create_evidence_from_rules(self, rule_results: List[Dict], verifications: Dict,
                                    extracted_fields: Dict, documents: List) -> List[EvidenceRecord]:
        evidence_list = []
        for rule in rule_results:
            if not rule.get("evidence_ref"):
                continue
            evidence = self.create_evidence({
                "requirement_id": rule.get("rule_id"),
                "document_id": rule.get("evidence_ref"),
                "page_number": self._get_page_number(rule, extracted_fields),
                "extracted_value": rule.get("actual", ""),
                "expected_value": rule.get("expected", ""),
                "actual_value": rule.get("actual", ""),
                "verification_source": rule.get("source", rule.get("rule_type", "")),
                "rule_id": rule.get("rule_id"),
                "result": rule.get("result"),
                "confidence": rule.get("confidence", 0),
                "details": rule.get("details", {})
            })
            evidence_list.append(evidence)
        return evidence_list

    def get_evidence(self, evidence_id: str) -> Optional[EvidenceRecord]:
        return self.evidence_store.get(evidence_id)

    def get_evidence_for_bid(self, bid_id: str) -> List[EvidenceRecord]:
        return [e for e in self.evidence_store.values()
                if e.details.get("bid_id") == bid_id or bid_id in e.requirement_id]

    def get_evidence_by_requirement(self, requirement_id: str) -> List[EvidenceRecord]:
        return [e for e in self.evidence_store.values() if e.requirement_id == requirement_id]

    def get_evidence_by_document(self, document_id: str) -> List[EvidenceRecord]:
        return [e for e in self.evidence_store.values() if e.document_id == document_id]

    def search_evidence(self, query: Dict[str, Any]) -> List[EvidenceRecord]:
        results = list(self.evidence_store.values())

        if query.get("result"):
            results = [e for e in results if e.result == query["result"]]
        if query.get("verification_source"):
            results = [e for e in results if e.verification_source == query["verification_source"]]
        if query.get("rule_id"):
            results = [e for e in results if e.rule_id == query["rule_id"]]
        if query.get("min_confidence"):
            results = [e for e in results if e.confidence >= query["min_confidence"]]
        if query.get("bid_id"):
            results = [e for e in results if e.details.get("bid_id") == query["bid_id"]]

        return sorted(results, key=lambda e: e.created_at, reverse=True)

    def add_override_evidence(self, params: Dict[str, Any]) -> EvidenceRecord:
        evidence = self.create_evidence({
            "requirement_id": params.get("requirement_id"),
            "document_id": f"OVERRIDE-{int(datetime.now().timestamp() * 1000)}",
            "page_number": 0,
            "extracted_value": params.get("original_result", ""),
            "expected_value": params.get("original_result", ""),
            "actual_value": params.get("new_result", ""),
            "verification_source": "MANUAL_OVERRIDE",
            "rule_id": params.get("rule_id"),
            "result": params.get("new_result"),
            "confidence": 1.0,
            "details": {
                "officer_id": params.get("officer_id"),
                "original_result": params.get("original_result"),
                "new_result": params.get("new_result"),
                "reason": params.get("reason"),
                "override": True,
                "timestamp": datetime.now(timezone.utc).isoformat()
            }
        })
        return evidence

    def get_evidence_summary(self, bid_id: str) -> Dict[str, Any]:
        evidence = self.get_evidence_for_bid(bid_id)
        by_result = {}
        by_source = {}
        for e in evidence:
            by_result[e.result] = by_result.get(e.result, 0) + 1
            by_source[e.verification_source] = by_source.get(e.verification_source, 0) + 1
        return {
            "bid_id": bid_id,
            "total_evidence": len(evidence),
            "by_result": by_result,
            "by_source": by_source,
            "evidence": [e.to_dict() for e in evidence]
        }

    def export_for_audit(self, bid_id: str) -> Dict[str, Any]:
        evidence = self.get_evidence_for_bid(bid_id)
        return {
            "bid_id": bid_id,
            "exported_at": datetime.now(timezone.utc).isoformat(),
            "total_records": len(evidence),
            "records": [e.to_dict() for e in evidence]
        }

    def _get_page_number(self, rule: Dict, fields: Dict) -> int:
        if rule.get("rule_id") == "REQ-LC-01":
            return fields.get("local_content_source_page", 1)
        if rule.get("rule_id") == "REQ-OEM-01":
            return fields.get("oem_source_page", 1)
        return rule.get("details", {}).get("source_page", 1)

    def clear(self):
        self.evidence_store.clear()


evidence_engine = EvidenceEngine()